src/repr/src/relation.rs

```diff
@@ -355,6 +359,65 @@ impl proptest::arbitrary::Arbitrary for ColumnName {
     }
 }
 
+/// Stable index of a column in a [`RelationDesc`].
+#[derive(
+    Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct ColumnIndex(usize);
+
+static_assertions::assert_not_impl_all!(ColumnIndex: Arbitrary);
+
+/// The version a given column was added at.
+#[derive(
+    Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct RelationVersion(u64);
+
+impl RelationVersion {
+    pub fn root() -> Self {
+        RelationVersion(0)
+    }
+
+    /// Returns an instance of [`RelationVersion`] which is "one" higher than `self`.
+    pub fn bump(&self) -> Self {
+        let next_version = self
+            .0
+            .checked_add(1)
+            .expect("added more than u64::MAX columns?");
+        RelationVersion(next_version)
+    }
+
+    /// Returns the inner value.
+    ///
+    /// TODO(parkmycar): Remove this [`RelationVersion`] should be opaque.
+    pub fn inner(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl RustType<ProtoRelationVersion> for RelationVersion {
+    fn into_proto(&self) -> ProtoRelationVersion {
+        ProtoRelationVersion { value: self.0 }
+    }
+
+    fn from_proto(proto: ProtoRelationVersion) -> Result<Self, TryFromProtoError> {
+        Ok(RelationVersion(proto.value))
+    }
+}
+
+/// Metadata (other than type) for a column in a [`RelationDesc`].
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
+struct ColumnMetadata {
```

Comment:
Questions about the data model:
- When do we set `ColumnMetadata::dropped`, and when do we remove the entry from the map entirely?
- At what point do we expect the caller to call `version_at`? Does that always happen by the time it's passed to Persist?

Reply:
(Sorry if these are redundant; I may be missing context!)

Reply:
Great questions!

> When do we set ColumnMetadata::dropped

We set `ColumnMetadata::dropped` when a user runs `ALTER TABLE ... DROP COLUMN ...`. We aren't planning to support this immediately but I figured while I was here I would add support.

> ...when do we remove the entry from the map entirely?

We won't ever drop the column entirely, but if a user runs `at_version(...)` with a version where the column doesn't exist, it won't be included in the returned `RelationDesc`s map.

> At what point do we expect the caller to call version_at? Does that always happen by the time it's passed to Persist?

This should get called during planning! Persist should never have to worry about calling `at_version(...)`. In fact, I put `at_version(...)` on a newtype so it shouldn't ever be possible to call `at_version(...)` outside of planning